### PR TITLE
Fixes #3841 'time_ago' & quantity plurals should be combined

### LINKED ISSUE #3841: 'time_ago' & quantity plurals should be combined
After #3795, TextViewBindingAdapters combines two strings (time_ago & plurals) into a single one. This ought to just be different plural options so that translators have better context when translating the full conceptual string.

### LINKED ISSUE #3842: Investigate second/millisecond conversion check for greeting timestamp
After #3795, TextViewBindingAdapters performs a check on whether the last visited timestamp is in milliseconds or seconds. We ought to just use milliseconds & never need this conversion, but investigation is needed to better understand where this check came from & if it could ever be triggered.

### PR DESCRIPTION
**Pull request by mbobiosio**
## Explanation

Fixes #3841
Fixes #3842
This PR includes the required update to combine strings together

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

File: app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java [hunk 1]
```diff
@@ -16,10 +16,14 @@
 import org.oppia.android.util.system.OppiaClock;
 import org.oppia.android.util.system.OppiaClockInjectorProvider;
 
-/** Holds all custom binding adapters that bind to [TextView]. */
+/**
+ * Holds all custom binding adapters that bind to [TextView].
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java:20]: Why this change? Per https://developer.android.com/kotlin/style-guide#formatting_2 the previous version is acceptable (and is what we use for such cases). While I linked to the KDoc one, the same also applies for Java: https://google.github.io/styleguide/javaguide.html#s7.1-javadoc-formatting.

Ditto for the others.

File: app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java [hunk 2]
```diff
@@ -107,24 +117,11 @@ private static String getPluralString(
       @PluralsRes int pluralsResId,
       int count
   ) {
-    // TODO(#3841): Combine these strings together.
-    return resourceHandler.getStringInLocaleWithWrapping(
-        R.string.time_ago,
-        resourceHandler.getQuantityStringInLocaleWithWrapping(
+    return resourceHandler.getQuantityStringInLocaleWithWrapping(
             pluralsResId, count, String.valueOf(count)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java:121]: Is this indentation still correct?

File: app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditFragmentPresenter.kt [hunk 3]
```diff
@@ -132,22 +133,21 @@ class ProfileEditFragmentPresenter @Inject constructor(
...
-              val intent =
-                Intent(fragment.requireContext(), AdministratorControlsActivity::class.java)
-              intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
-              fragment.startActivity(intent)
-            } else {
-              val intent = Intent(fragment.requireContext(), ProfileListActivity::class.java)
-              intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
-              fragment.startActivity(intent)
-            }
+        fragment
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/settings/profile/ProfileEditFragmentPresenter.kt:136]: I'm not opposed to including this & other cleanup changes in this PR, but something to keep in mind in the future is that it's not always best practice to include unrelated changes in a PR to the central thing the PR is trying to accomplish. Doing so actually introduces some risk in that it can cause the PR to need to be reverted if the refactor has an unexpected behavior change, or it can be difficult to revert the PR due to a larger number of changes than what's actually necessary to accomplish its goal.

File: app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt [hunk 4]
```diff
@@ -30,4 +31,49 @@ class DateTimeUtil @Inject constructor(
...
+        diff.diffToString(WEEKS)
+      )
+      else -> resourceHandler.getQuantityStringInLocaleWithWrapping(
+        R.plurals.years_ago, diff.diffToInt(MONTHS), diff.diffToString(MONTHS)
+      )
+    }
+  }
+
+  companion object {
+    fun currentDate() = Calendar.getInstance().timeInMillis
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt:67]: Why is this being used vs. passing in the current time? I think we conventionally use OppiaClock for calculating the current time, so I'm trying to understand this difference.

File: utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt [hunk 5]
```diff
@@ -1,6 +1,7 @@
-package org.oppia.android.app.utility
+package org.oppia.android.util.extensions
 
 import android.content.Context
+import android.widget.Toast
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt:4]: Unused import--please remove.

File: app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt [hunk 6]
```diff
@@ -116,7 +116,9 @@ class AppVersionActivityTest {
 
   @get:Rule
   val activityTestRule: ActivityTestRule<AppVersionActivity> = ActivityTestRule(
-    AppVersionActivity::class.java, /* initialTouchMode= */ true, /* launchActivity= */ false
+    AppVersionActivity::class.java, /* initialTouchMode= */
+    true, /* launchActivity= */
+    false
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt:121]: This formatting became a bit messed up--please fix.

File: app/src/main/res/values/plurals.xml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
+  <plurals name="minutes_ago">
```
Comment (BenHenning) [inline app/src/main/res/values/plurals.xml:4]: We usually define plurals in strings.xml--is there a reason these can't go there?

Comment (BenHenning) [inline app/src/main/res/values/plurals.xml:4]: For this & the others: the plural name should more clearly indicate what this is being used for as that provides better context to translators and within code. For example: ``last_logged_in_minutes_ago`` might be a much clearer alternative.

File: app/src/main/res/values/plurals.xml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    <item quantity="one">%s month ago</item>
+    <item quantity="other">%s months ago</item>
+  </plurals>
+
+  <plurals name="years_ago">
+    <item quantity="one">%s year ago</item>
+    <item quantity="other">%s years ago</item>
+  </plurals>
+
+</resources>
```
Comment (BenHenning) [inline app/src/main/res/values/plurals.xml:34]: Nit: all files should end with a single EOF newline.

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -342,6 +342,7 @@
   <string name="profile_edit_delete_dialog_title">Permanently delete this profile?</string>
   <string name="profile_edit_delete_dialog_message">All progress will be deleted and cannot be recovered.</string>
   <string name="profile_edit_delete_dialog_positive">Delete</string>
+  <string name="profile_edit_delete_successful_message">Profile deleted successfully</string>
```
Comment (BenHenning) [inline app/src/main/res/values/strings.xml:345]: This seems unrelated to your change--why is it needed?

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -467,6 +468,16 @@
...
+
+  <string name="minute_ago">a minute ago</string>
+  <string name="minutes_ago">%s minutes ago</string>
+  <string name="hour_ago">an hour ago</string>
+  <string name="hours_ago">%s hours ago</string>
+  <string name="days_ago">%s days ago</string>
+  <string name="month_ago">a month ago</string>
+  <string name="months_ago">%s months ago</string>
+  <string name="year_ago">a year ago</string>
+  <string name="years_ago">%s years ago</string>
```
Comment (BenHenning) [inline app/src/main/res/values/strings.xml:480]: Are these still needed? It seems like your change is only using the new plurals.

Also, are there any old strings that can now be removed in favor of the new solution?

File: app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt [hunk 11]
```diff
@@ -30,4 +31,49 @@ class DateTimeUtil @Inject constructor(
     /** Returns [DateTimeUtil] for the current Dagger graph. */
     fun getDateTimeUtil(): DateTimeUtil
   }
+
+  /**
+   * Returns the readable string of the duration from the provided time in [Long].
+   */
+  fun timeAgoFromTimestamp(timestamp: Long, referenceTime: Long): String {
+    val diff = (currentDate() - timestamp) / MILLI_SECONDS
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt:39]: For this & the constants below, could we use ``TimeUnit``, instead? That both avoids having to do time math (which can be easily misunderstood and/or written), and may help improve readability. An example of what this might look like:

```kotlin
val diffSeconds = TimeUnit.MILLISECONDS.toSeconds(currentTimeMillis - timestampMillis)
return when {
  diffSeconds < TimeUnit.MINUTES.toSeconds(1) -> ...
  diffSeconds < TimeUnit.HOURS.toSeconds(1) -> ...
  ...
}
```

I think ``TimeUnit`` could also simplify your ``diffTo*`` functions by providing easier conversions. There might even be a way to generalize this slightly, e.g. something like:

```kotlin
private val timestampFormatters by lazy {
  listOf(
    TimestampFormatter(threshold = TimeMoment.ONE_MINUTE, timeUnit = TimeUnit.SECONDS, formatTimestamp = this::formatSecondsTimestamp),
    TimestampFormatter(threshold = TimeMoment.ONE_HOUR, timeUnit = TimeUnit.MINUTES, formatTimestamp = this::formatMinutesTimestamp),
    ...
  )
}

fun formatTimeAgoTimestamp(timestampMillis: Long): String {
  val timestamp = TimeMoment(amount = timestampMillis, unit = TimeUnit.MILLISECONDS)
  val currentTime = TimeMoment(amount = oppiaClock.getCurrentTimeMillis(), unit = TimeUnit.MILLISECONDS)
  val timeSpent = currentTime - timestamp
  return timestampFormatters.firstOrNull {
    it.canFormatWithinThreshold(timeSpent)
  }?.formatString(timeSpent) ?: error("Encountered impossible time: $timeSpent (for time: $timestamp)")
}

private class TimestampFormatter(val threshold: TimeSpan, val timeUnit: TimeUnit, val formatTimestamp: (TimeMoment) -> String) {
  fun canFormatWithinThreshold(timestamp: TimeMoment): Boolean = timestamp.isEarlierThan(threshold)

  fun format(timestamp: TimeMoment): String = formatTimestamp(timestamp.convertTo(timeUnit))
}

private fun formatSecondsTimestamp(seconds: TimeMoment): String =
  resourceHandler.getStringInLocale(R.string.just_now)

private fun formatMinutesTimestamp(minutes: TimeMoment): String {
  return resourceHandler.getQuantityStringInLocaleWithWrapping(
    R.plurals.minutes_ago, minutes.asInt, minutes.asString
  )
}
...

// A simple alternative for Java 8-only time utilities.
private data class TimeMoment(val amount: Long, val unit: TimeUnit) {
  val inMilliseconds: Long
    get() = TimeUnit.MILLISECONDS.convert(amount, unit)

  val asInt: Int
    get() = amount.toInt()

  val asString: String
    get() = amount.toString()

  fun isEarlierThan(other: TimeMoment): Boolean = inMilliseconds < other.inMilliseconds

  fun convertTo(unit: TimeUnit): TimeMoment = TimeMoment(unit.convert(amount, this.unit), unit)

  operator fun minus(other: TimeMoment): TimeMoment =
    TimeMoment(inMilliseconds - other.inMilliseconds, TimeUnit.MILLISECONDS)

  companion object {
    val ONE_MINUTE = TimeMoment(amount = 1, unit = TimeUnit.MINUTES)
    val ONE_HOUR = TimeMoment(amount = 1, unit = TimeUnit.HOURS)
    ...
  }
}
```

(As this is just a concept, more work & reformatting is needed).

File: app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt [hunk 12]
```diff
@@ -30,4 +31,49 @@ class DateTimeUtil @Inject constructor(
     /** Returns [DateTimeUtil] for the current Dagger graph. */
     fun getDateTimeUtil(): DateTimeUtil
   }
+
+  /**
+   * Returns the readable string of the duration from the provided time in [Long].
+   */
+  fun timeAgoFromTimestamp(timestamp: Long, referenceTime: Long): String {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt:38]: Method names should be verbs or verb phrases describing what's happening (e.g. formatting, or converting, in this case).

Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/utility/datetime/DateTimeUtil.kt:38]: Please add new tests for this method to cover all possible pluralizations that should now be supported.

File: app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import android.app.Activity
+import android.content.ContextWrapper
+import android.view.View
+import android.widget.TextView
+import androidx.databinding.BindingAdapter
+import org.oppia.android.R
+import org.oppia.android.app.translation.AppLanguageActivityInjectorProvider
+import org.oppia.android.app.translation.AppLanguageResourceHandler
+
+@BindingAdapter("profile:lastVisited")
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapter.kt:12]: Are these new binding adapters needed for this PR?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @mbobiosio. I took a first pass and had some comments--PTAL.

Also, CI checks are failing--please fix.

File: app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java (hunk 1 above)
Comment (mbobiosio) [reply app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java:20]: Corrected this

File: app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java (hunk 2 above)
Comment (mbobiosio) [reply app/src/main/java/org/oppia/android/app/databinding/TextViewBindingAdapters.java:121]: Yes this is correct
